src/arkode/arkode_lsrkstep_io.c

```diff
@@ -0,0 +1,661 @@
+{
+  ARKodeMem ark_mem;
+  ARKodeLSRKStepMem step_mem;
+  int retval;
+
+  /* access ARKodeMem and ARKodeLSRKStepMem structures */
+  retval = lsrkStep_AccessARKODEStepMem(arkode_mem, __func__, &ark_mem,
+                                        &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* get values from step_mem */
+  *avg_stage = ((sunrealtype)step_mem->nfe) /
+               ((sunrealtype)ark_mem->nst_attempts);
+
+  return (ARK_SUCCESS);
+}
+
+/*===============================================================
+  Private functions attached to ARKODE
+  ===============================================================*/
+
+/*---------------------------------------------------------------
+  lsrkStep_SetDefaults:
+
+  Resets all LSRKStep optional inputs to their default values.
+  Does not change problem-defining function pointers or
+  user_data pointer.
+  ---------------------------------------------------------------*/
+int lsrkStep_SetDefaults(ARKodeMem ark_mem)
+{
+  ARKodeLSRKStepMem step_mem;
+  int retval;
+  long int lenrw, leniw;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* Remove current SUNAdaptController object, and replace with "PI" */
+  if (ark_mem->hadapt_mem->owncontroller)
+  {
+    retval = SUNAdaptController_Space(ark_mem->hadapt_mem->hcontroller, &lenrw,
+                                      &leniw);
+    if (retval == SUN_SUCCESS)
+    {
+      ark_mem->liw -= leniw;
+      ark_mem->lrw -= lenrw;
+    }
+    retval = SUNAdaptController_Destroy(ark_mem->hadapt_mem->hcontroller);
+    ark_mem->hadapt_mem->owncontroller = SUNFALSE;
+    if (retval != SUN_SUCCESS)
+    {
+      arkProcessError(ark_mem, ARK_MEM_FAIL, __LINE__, __func__, __FILE__,
+                      "SUNAdaptController_Destroy failure");
+      return (ARK_MEM_FAIL);
+    }
+  }
+  ark_mem->hadapt_mem->hcontroller = SUNAdaptController_PI(ark_mem->sunctx);
+  if (ark_mem->hadapt_mem->hcontroller == NULL)
+  {
+    arkProcessError(ark_mem, ARK_MEM_FAIL, __LINE__, __func__, __FILE__,
+                    "SUNAdaptControllerPI allocation failure");
+    return (ARK_MEM_FAIL);
+  }
+  ark_mem->hadapt_mem->owncontroller = SUNTRUE;
+  retval = SUNAdaptController_Space(ark_mem->hadapt_mem->hcontroller, &lenrw,
+                                    &leniw);
+  if (retval == SUN_SUCCESS)
+  {
+    ark_mem->liw += leniw;
+    ark_mem->lrw += lenrw;
+  }
+
+  /* Set default values for integrator optional inputs
+     (overwrite some adaptivity params for LSRKStep use) */
+  step_mem->req_stages = 0; /* no stages */
+
+  /* Spectral info */
+  step_mem->lambdaR      = 0;
+  step_mem->lambdaI      = 0;
+  step_mem->sprad        = 0;
+  step_mem->spr_max      = 0;
+  step_mem->spr_min      = 0;
+  step_mem->dom_eig_sfty = 1.01;
+  step_mem->dom_eig_freq = 25;
```

Comment:
How was `25` selected as the default?

Reply:
As with most of our heuristic parameters, the "optimal" value is likely problem-dependent.  However, we try to never require users to change these values directly, instead opting for a reasonable default value.

@maggul: what was the default value in the older Fortran "RKC" code that you looked through?  I'm guessing it was where "25" came from.  Given the authors of that code, I anticipate that they tested it rather thoroughly, and so it would be a reasonable default for us (we could even add a citation for where our default came from).

Alternately, we could use "20" since that is the default of our related heuristic parameter for the number of steps between Jacobian/preconditioner updates in ARKStep.

Reply:
It is given in the last sentence of the 3.3 in [this](https://pdf.sciencedirectassets.com/271610/1-s2.0-S0377042700X00532/1-s2.0-S0377042797002197/main.pdf?X-Amz-Security-Token=IQoJb3JpZ2luX2VjEGoaCXVzLWVhc3QtMSJIMEYCIQDpKctTqmNVxFiSGcRVsRkHYzoh5iDmXtioGUYhgenIhAIhAKR3TAly%2F9PVO1fZxoEQJAV3TYzLA1%2FOXMaoM9xfbzoxKrsFCLL%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBRoMMDU5MDAzNTQ2ODY1IgyBJHSn6Hk7%2BXa5RvMqjwUjGI4XZRJ5R0ZjMgKx6kui6U8TG6%2FiRvwopm8GmoRCmpGTDutMKweban4UDg2vKW3wDP8hhRotTD0Mxv6g9gKZBiASlgfSFiMcGcju5n%2B8N%2FE5pyFkJvFLPrVGzfiPwJi3w1HJ3KgmtOhN0wCje2Y2EDQxOuTmF02sLXqx02GXjaONr7A1RRU%2FWwNlEggxIKzRJ3A6dt0lHopgyX2zOGGHTejmgQBpYjlLVGyfMNBBiAfOD0NzWA4iKr3ADgwgoBSrfV9BKmpKPtG7ZpLBbb1jqFKA6BIEljEUUHKVigL0sO8EidOpeNnjpstJY%2FFWJxUx2tAPtbfZY3tesb0VR6dnSq9QlXvF4NtC8gWmOyVjhSr%2BNNp8dspy2WAL1LOj5PQi8bJc%2BRoIeRVn6bjp2R%2B2fudiEkbFvvU2BgoFPcXS4PCjiYEgthPY1qpPZI0KDmeAWUO1ZB8mMWGMDQR4MAZRp6xH9yGz2iS60XMA6KcrrZpiSiv3TiH3zh55DJaoag5%2FD%2B7sd8hOgawagXXDBx72CgYWOtSgEVgV3jPxsqs0hKZieA8J%2FjbwWvhaOR3r2LCTY%2FtGqHN6yGCe3Le9qoT939K6xLejUHecP01m7bl6tMlIibZ6ZeuU8OXNPQubAQpFWAeFo8ga4UvtQtyxEDWnfwJxlxyfQTiKXyUkxmYWgO1hIQHSdJJc7qGV4KO%2FWrz0LQzrdGk2jCN2wTxcrzTVSCFGttwLEaV5OekNoLDhosDsZfnjIWFHEg1NFCUAX6fGDbuXhA6AAhXNmE1EGcOnS0XPSlM5%2B43t0NtfxYH4k%2BRkz7rNNBpGrn9msYmA31oelfTbPg20OCLSZJrdwsJLcz%2Bl2eYRNSPSMDATk9XVMPvd97cGOrAB9xVbC3zR7RJYaxVQY9%2BPynKTZh91eCFQAe5tnR%2FJnkSrv3X7mKP4lYXZm6QwDfVcpvNSzddYf8eOYR9QnTIiFxe33ae5uZa4eDMZkwyEOq0dbvq7Lr3e7oHjEQF9M461Jp6R9vrE%2BEAQvnHXYRgZq4Qxu2POgKfs1SqxyF0iVBe%2BWv15RB3IBNKkNw47Z30U%2B%2FvP0oXUj1xRiSU1ytODIJ6jv2FP3BAvK%2FV9CkXRci8%3D&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20241003T014853Z&X-Amz-SignedHeaders=host&X-Amz-Expires=300&X-Amz-Credential=ASIAQ3PHCVTYYZQZFD4N%2F20241003%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=407e4bd7d0af4f6b52ad5a1174d5ad90d5036c7f2f8bc7f111297cda2c2d1060&hash=9942dcd321816a7912b426985e951d48cb7bc1ce6402b1f7a0dc625a3e962ed8&host=68042c943591013ac2b2430a89b270f6af2c76d8dfd086a07176afe7c76c2c61&pii=S0377042797002197&tid=spdf-bc326f2e-ca9b-4970-99d3-52a9d65791cb&sid=56ac56ee972cd1442b18d7a434d1de8c066dgxrqa&type=client&tsoh=d3d3LnNjaWVuY2VkaXJlY3QuY29t&ua=131d5800075c55550353&rr=8cc945c12b2c6b65&cc=us).

Reply:
In that case, I think that we should add the following `.. note::` to the documentation for `LSRKStepSetDomEigFrequency`:

```
.. note::  The default value is 25 (as recommended in :cite:p:`<insert new citation to https://doi.org/10.1016/S0377-0427(97)00219-7>`).  An input ``nsteps < 0`` will reset to this default.
```
And you should change the implementation to have a negative input reset to the default instead of returning an error.